### PR TITLE
Replace hard-coded paths with environment variables

### LINKED ISSUE #2272: tiles version hardcoded to trunk
The version of the tiles product is hardcoded to trunk in `desispec.io.fibermap.assemble_fibermap()` [L595](https://github.com/desihub/desispec/blob/884969a3298370f97a8409605571a6b623e99977/py/desispec/io/fibermap.py#L595), and possibly in other places. This means there is no possible way to specify a tag of the tiles product when creating a specprod for release, *e.g.* `jura`.

The tag is important for creating a fixed set of fiberassign files that will be:

1. Included in a public release
2. Themselves loaded into the specprod database
3. The definition of the observed and potential targets for the photometry catalog associated with a specprod, which will also be loaded into the specprod database

We had a bit of good luck though: we can easily reconstruct the revision of trunk used for `jura` and create a tag based on that (which will be 2.0). If changes to the fiberassign files are needed for `kilimanjaro`, that could be tag 2.1 (but hopefully that won't be necessary).

We still need to establish whether the line above is the only place where the tiles version is hardcoded. I will attempt to do a very large `grep`, since this problem may not be limited to `desispec`.

For future runs, what is the best way to pass information on the tiles tag to the pipeline? One possibility is to associate a "tiles" modulefile that points to the existing tiles checkout (trunk or a tag) on CFS, and then pass that information via an environment variable.  Or the environment variable could be set in another modulefile like desimodules (see also desihub/desimodules#49). If the environment variable is not set, that should be considered an error, rather than falling back on a hardcoded value.

@sbailey @araichoor @akremin @geordie666 *et al.* please take note. Tag others as needed.

### PR DESCRIPTION
**Pull request by weaverba137**
This PR fixes #2272.

This PR introduces a new environment variable, `DESI_TILES`, similar to `DESI_SURVEYOPS` and also uses `DESI_SURVEYOPS` more extensively.

`DESI_TILES` will be defined in desimodules (as a fix to desihub/desimodules#49).

### CODE REVIEW

**Comment by sbailey** [comment]
To emphasize the WIP status as a warning to my future self: this requires updates to desimodules to define $DESI_TILES before it will work, and there are other places that also reference trunk that have not yet been updated (e.g. desispec.io.fibermap.assemble_fibermap).  But in the spirit of WIP, the basic structure of `get_readonly_filepath(os.environ['DESI_BLATFOO'])` shown here looks good.

**Comment by weaverba137** [comment]
@sbailey, while working on this, I noticed that fiberassign file names are specified in multiple ways *e.g.* `f'{tileid // 1000:03d}` versus `stileid = f'{tileid:06d}'; ... stileid[:3] ...`. Should we define a function instead and only ever use the function to find the fiberassign file?

**Comment by sbailey** [comment]
> @sbailey, while working on this, I noticed that fiberassign file names are specified in multiple ways e.g. f'{tileid // 1000:03d} versus stileid = f'{tileid:06d}'; ... stileid[:3] .... Should we define a function instead and only ever use the function to find the fiberassign file?

Good point.  Best would be to add a new filetype entry to `desispec.io.findfile`, e.g. "fiberassignsvn", which would then do this $DESI_TILE+tileid etc. wrangling to report the canonical name of a fiberassign file in svn.  This would augment the pre-existing filetype='fiberassign' which reports where the fiberassign file is in the raw data stream.

**Comment by weaverba137** [comment]
OK, and while working on `findfile`, I found this. This can't be right can it (around line 294)?
```python
    if isinstance(tile, str):    tile = int(night)
```

**Comment by sbailey** [comment]
Indeed, that whole block was introduced (by me!) in PR #1931 and is a cut-and-paste typo on multiple lines:
```python
    #- be robust to str vs. int
    if isinstance(healpix, str): healpix = int(healpix)
    if isinstance(night, str):   night = int(night)
    if isinstance(expid, str):   expid = int(night)
    if isinstance(tile, str):    tile = int(night)
    if isinstance(spectrograph, str): spectrogrpah = int(night)
```
Note the "spectrogrpah" typo as well.

@weaverba137 would you like me to fix these in a separate PR (with tests!) or will you fix it in your current branch?

**Comment by weaverba137** [comment]
I will do a minimal fix on this branch, but let's think about a separate PR to do more robust unit tests on `findfile`.

In addition, I think a future policy for all PRs should be "if you are trying to create a file name, you **must** use `findfile`".

**Comment by weaverba137** [comment]
I'm moving this out of draft status.

@sbailey, please review [this listing of 'trunk' occurrences](https://github.com/search?q=repo%3Adesihub%2Fdesispec+trunk+NOT+trunk_py+NOT+language%3A%22Jupyter+Notebook%22+NOT+path%3Adoc&type=code). I believe these files can be ignored:

1. etc/cron_dailytest.sh
2. etc/desi_nersc_pipetest.py
3. py/desispec/database/redshift.py (the entire database directory is deprecated anyway)
4. py/desispec/test/test_photo.py
5. py/desispec/test/integration_test.py

I'm not sure about these, because a URL is being constructed instead of a filename:

6. py/desispec/scripts/procdashboard.py
7. py/desispec/scripts/zprocdashboard.py

These files have been updated:

8. py/desispec/io/photo.py
9. bin/desi_tile_qa_reference
10. py/desispec/io/fibermap.py
11. py/desispec/workflow/exptable.py
12. py/desispec/scripts/archive_night.py

File: py/desispec/io/fibermap.py [hunk 1]
```diff
@@ -590,20 +591,16 @@ def assemble_fibermap(night, expid, badamps=None, badfibers_filename=None,
...
-        try:
-            fafile = checkgzip(testfile)
-        except FileNotFoundError:
-            # no alternate fiberassign file in svn yet, that's ok
-            pass
-
-        if rawfafile != fafile:
-            log.info(f'Overriding raw fiberassign file {rawfafile} with svn {fafile}')
+    if allow_svn_override:
+        if 'DESI_TILES' in os.environ:
```
Comment (weaverba137) [inline py/desispec/io/fibermap.py:595]: Technically, this violates our proposed rule of 'if `DESI_TILES` is not defined, fail', so I can make this stricter if necessary.

**Comment by akremin** [comment]
The procdashboards are used frequently and so a refactor is welcome, but I
also think it is okay to leave as-is since it is a special case where it
constructs a url. If you want, you are welcome to update the logic so that
it uses findfile for the relative path and then creates the url based on
that. But I'm also fine with leaving it. These files aren't moving anytime
soon, so keeping the dashboard code as-is is very low risk.

**Review by sbailey — CHANGES_REQUESTED** [review]

File: py/desispec/workflow/exptable.py [hunk 2]
```diff
@@ -692,21 +693,14 @@ def summarize_exposure(raw_data_dir, night, exp, obstypes=None, colnames=None, c
...
-                                    f'fiberassign-{tileid:06d}.fits')
+            fbasvn, svn_exists = findfile('fiberassignsvn',
+                                          tile=tileid, return_exists=True)
 
             fbafinal = fbaraw
-            if os.path.exists(fbasvn):
-                fbafinal = fbasvn
-            elif os.path.exists(fbasvn+'.gz'):
-                fbasvn = fbasvn+'.gz'
+            if svn_exists:
```
[the remark this answers is not in the document]
Comment (weaverba137) [reply py/desispec/workflow/exptable.py:703]: Ah, you're just making a comment here, not actually requesting changes?

File: py/desispec/workflow/exptable.py [hunk 3]
```diff
@@ -692,21 +693,14 @@ def summarize_exposure(raw_data_dir, night, exp, obstypes=None, colnames=None, c
...
-                fbaraw = fbaraw+'.gz'
+            fbaraw, raw_exists = findfile('fiberassign', night=night, expid=expstr,
+                                          tile=tileid, return_exists=True)
 
-            targdir = os.getenv('DESI_TARGET')
-            fbasvn = os.path.join(targdir, 'fiberassign', 'tiles', 'trunk',
-                                    f'{tileid // 1000:03d}',
-                                    f'fiberassign-{tileid:06d}.fits')
+            fbasvn, svn_exists = findfile('fiberassignsvn',
+                                          tile=tileid, return_exists=True)
```
[the remark this answers is not in the document]
Comment (weaverba137) [reply py/desispec/workflow/exptable.py:700]: Done.

File: py/desispec/io/photo.py [hunk 4]
```diff
@@ -36,27 +36,28 @@ def gather_targetdirs(tileid, fiberassign_dir=None, verbose=False):
...
-        fiberfile = fiberfile.replace('.gz', '')
-        if not os.path.isfile(fiberfile):
-            errmsg = f'Fiber assignment file {fiberfile} not found!'
-            log.critical(errmsg)
-            raise IOError(errmsg)
+        fiberassign_dir = get_readonly_filepath(os.environ['DESI_TILES'])
+        # fiberassign_dir = os.path.join(desi_root, 'target', 'fiberassign', 'tiles', 'trunk')
+
+    fiberfile, svn_exists = findfile('fiberassignsvn', tiles_dir=fiberassign_dir,
+                                     tile=tileid, readonly=True, return_exists=True)
```
[the remark this answers is not in the document]
Comment (weaverba137) [reply py/desispec/io/photo.py:56]: I think what you're saying is to just remove the `tiles_dir` argument. Is that correct?

**Comment by weaverba137** [comment]
I can't add comments directly to this line item:

> Thanks for noting that. Seeing that you've added a tiles_dir option to findfile, this code could intercept missing $DESI_TILES and fall back to doing what it would have done before, i.e. using tiles_dir=$DESI_TARGET/fiberassign/tiles/trunk instead (while still printing a warning).

I'm not entirely sure what you're recommending here. Maybe you could express this as a code snippet?

However, I now realize that the way `findfile` is modified in this branch, an exception will be thrown if `FIBER_ASSIGN_DIR` is not defined *anyway*, and so maybe the solution is to catch a potential `KeyError` thrown by findfile, add a helpful message, and re-raise.

File: py/desispec/workflow/exptable.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbailey) [reply py/desispec/workflow/exptable.py:703]: yes, this was just for the record in the PR; no change needed here.

File: py/desispec/io/photo.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbailey) [reply py/desispec/io/photo.py:56]: yes

**Comment by sbailey** [comment]
> I'm not entirely sure what you're recommending here. Maybe you could express this as a code snippet?

Thinking about it more, let's just go all-in on making $FIBER_ASSIGN_DIR a required variable for standard operations, and not try to support an interim period where it might not be defined, since desimodules/main already defines it.

Heads up @sybenzvi, when you update desispec at KPNO, you'll also need to make sure that module files / config gets set to define FIBER_ASSIGN_DIR=$DESI_TARGET/fiberassign/tiles/trunk .

> However, I now realize that the way findfile is modified in this branch, an exception will be thrown if FIBER_ASSIGN_DIR is not defined anyway, and so maybe the solution is to catch a potential KeyError thrown by findfile, add a helpful message, and re-raise.

We're ok here.  You added
```python
if tiles_dir is None and 'tiles_dir' in required_inputs:
        tiles_dir = os.environ['FIBER_ASSIGN_DIR']
```
which means that $FIBER_ASSIGN_DIR is only required if tiles_dir is *not* specified, which is good.  We'll require $FIBER_ASSIGN_DIR for normal operations, but if for whatever reason someone is running without it, `findfile('fiberassignsvn', tiles_dir=...)` will work without requiring an envvar that it doesn't use (I explicitly checked that).

Summarizing, I think we want the `assemble_fibermap` block to be
```python
    if allow_svn_override:
        testfile, svn_exists = findfile('fiberassignsvn', tile=tileid, readonly=True, return_exists=True)
        if svn_exists:
            fafile = testfile
            log.info(f'Overriding raw fiberassign file {rawfafile} with svn {fafile}')
        else:
            log.info(f'{testfile}[.gz] not found; sticking with raw data fiberassign file')
```
i.e. findfile will require $FIBER_ASSIGN_DIR to exist and raise a KeyError if it doesn't, and this code handles the case where findfile works but the tile hasn't yet been committed to svn (svn_exists=False).

File: py/desispec/io/photo.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (weaverba137) [reply py/desispec/io/photo.py:56]: OK, that also means that the argument to `desispec.io.photo.gather_targetdirs()` is now unused. Should I remove the argument?

[the remark this answers is not in the document]
Comment (sbailey) [reply py/desispec/io/photo.py]: Sorry, I had missed that it is also an input argument, not just a temporary variable constructed from environment variables.  Updated suggestion: just use
```
fiberfile, svn_exists = findfile('fiberassignsvn', tiles_dir=fiberassign_dir,
                                     tile=tileid, readonly=True, return_exists=True)
```
without any pre-wrangling of `fiberassign_dir`.  If the input is `None`, then findfile will get it from $FIBER_ASSIGN_DIR, and if the input is not None, then findfile will use that.  No need for `gather_targetdirs` to do its own manipulation of `if fiberassign_dir is None`.

If that doesn't make sense, ping me on Slack and let's hop on zoom to sort this out in realtime.

**Comment by weaverba137** [comment]
@sbailey, I've made the requested changes, except for the open question about `io.photo` above.

File: py/desispec/io/photo.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (weaverba137) [reply py/desispec/io/photo.py]: Thanx, we cross posted about this, but this is a good suggestion.

[the remark this answers is not in the document]
Comment (weaverba137) [reply py/desispec/io/photo.py:52]: The most recent commit should have all the requested changes in now.

**Review by sbailey — APPROVED** [review]
Looks good!  Thanks for iterating on all the details.

**Comment by sbailey** [comment]
For the record: I have installed this updated version at NERSC

**Pull request merged**